File: shared/src/lib/eth_consensus_layer.rs
use ethereum_types::{H160, H256, U256};
use serde::{Deserialize, Serialize};
use ssz_derive::{Decode, Encode};
pub use ssz_types::{typenum, typenum::Unsigned, BitList, BitVector, FixedVector, VariableList};

use tree_hash::TreeHash;
use tree_hash_derive::TreeHash;

pub type Address = H160;
pub type CommitteeIndex = u64;
pub type Hash256 = H256;
pub type Root = Hash256;
pub type BlsPublicKey = FixedVector<u8, typenum::U48>;
pub type ForkVersion = FixedVector<u8, typenum::U4>;
pub type Version = FixedVector<u8, typenum::U4>;
pub type ParticipationFlags = u8;

use crate::eth_spec;

type Slot = u64;
type Epoch = u64;

// Re-export
pub type SlotsPerEpoch = eth_spec::SlotsPerEpoch;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct Fork {
    previous_version: Version,
    current_version: Version,
    epoch: Epoch,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct Checkpoint {
    epoch: Epoch,
    root: Root,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct BeaconBlockHeader {
    pub slot: Slot,
    pub proposer_index: CommitteeIndex,
    pub parent_root: Root,
    pub state_root: Root,
    pub body_root: Root,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct Eth1Data {
    deposit_root: Root,
    // #[serde(with = "serde_utils::quoted_u64")]
    deposit_count: u64,
    block_hash: Hash256,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct Validator {
    pub pubkey: BlsPublicKey,
    pub withdrawal_credentials: Hash256,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub effective_balance: u64,
    pub slashed: bool,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub activation_eligibility_epoch: Epoch,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub activation_epoch: Epoch,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub exit_epoch: Epoch,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub withdrawable_epoch: Epoch,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct AttestationData {
    pub slot: Slot,
    pub index: CommitteeIndex,
    pub beacon_block_root: Root,
    pub source: Checkpoint,
    pub target: Checkpoint,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct PendingAttestation {
    aggregation_bits: BitList<eth_spec::MaxValidatorsPerCommittee>,
    data: AttestationData,
    inclusion_delay: Slot,
    proposer_index: CommitteeIndex,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct SyncCommittee {
    pubkeys: FixedVector<BlsPublicKey, eth_spec::SyncCommitteeSize>,
    aggregate_pubkey: BlsPublicKey,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct ExecutionPayloadHeader {
    parent_hash: Hash256,
    fee_recipient: Address,
    state_root: Root,
    receipts_root: Root,
    logs_bloom: FixedVector<u8, eth_spec::BytesPerLogBloom>,
    prev_randao: Hash256,
    // #[serde(with = "serde_utils::quoted_u64")]
    block_number: u64,
    // #[serde(with = "serde_utils::quoted_u64")]
    gas_limit: u64,
    // #[serde(with = "serde_utils::quoted_u64")]
    gas_used: u64,
    // #[serde(with = "serde_utils::quoted_u64")]
    timestamp: u64,
    extra_data: VariableList<u8, eth_spec::MaxExtraDataBytes>,
    // workaround - looks like ByteList is partially broken, but extra data is exactly bytes32
    // extra_data: Hash256
    base_fee_per_gas: U256,
    block_hash: Hash256,
    transactions_root: Root,
    withdrawals_root: Root,
    // excess_data_gas: Uint256
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct HistoricalSummary {
    block_summary_root: Root,
    state_summary_root: Root,
}

pub type Validators = VariableList<Validator, eth_spec::ValidatorRegistryLimit>;
pub type Balances = VariableList<u64, eth_spec::ValidatorRegistryLimit>;

// Simplified https://github.com/sigp/lighthouse/blob/master/consensus/types/src/beacon_state.rs#L212
// Primarily - flattening the "superstruct" part on different eth specs,
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct BeaconState {
    // Versioning
    // #[serde(with = "serde_utils::quoted_u64")]
    pub genesis_time: u64,
    pub genesis_validators_root: Hash256,
    pub slot: Slot,
    pub fork: Fork,

    // History
    pub latest_block_header: BeaconBlockHeader,
    pub block_roots: FixedVector<Hash256, eth_spec::SlotsPerHistoricalRoot>,
    pub state_roots: FixedVector<Hash256, eth_spec::SlotsPerHistoricalRoot>,
    // Frozen in Capella, replaced by historical_summaries
    pub historical_roots: VariableList<Hash256, eth_spec::HistoricalRootsLimit>,

    // Ethereum 1.0 chain data
    pub eth1_data: Eth1Data,
    pub eth1_data_votes: VariableList<Eth1Data, eth_spec::SlotsPerEth1VotingPeriod>,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub eth1_deposit_index: u64,

    // Registry
    pub validators: Validators,
    // #[serde(with = "ssz_types::serde_utils::quoted_u64_var_list")]
    pub balances: Balances,

    // Randomness
    pub randao_mixes: FixedVector<Hash256, eth_spec::EpochsPerHistoricalVector>,

    // Slashings
    // #[serde(with = "ssz_types::serde_utils::quoted_u64_fixed_vec")]
    pub slashings: FixedVector<u64, eth_spec::EpochsPerSlashingsVector>,

    // Participation (Altair and later)
    pub previous_epoch_participation: VariableList<ParticipationFlags, eth_spec::ValidatorRegistryLimit>,
    pub current_epoch_participation: VariableList<ParticipationFlags, eth_spec::ValidatorRegistryLimit>,

    // Finality
    pub justification_bits: BitVector<eth_spec::JustificationBitsLength>,
    pub previous_justified_checkpoint: Checkpoint,
    pub current_justified_checkpoint: Checkpoint,
    pub finalized_checkpoint: Checkpoint,

    // Inactivity
    // #[serde(with = "ssz_types::serde_utils::quoted_u64_var_list")]
    pub inactivity_scores: VariableList<u64, eth_spec::ValidatorRegistryLimit>,

    // Light-client sync committees
    pub current_sync_committee: SyncCommittee,
    pub next_sync_committee: SyncCommittee,

    // Execution
    pub latest_execution_payload_header: ExecutionPayloadHeader,

    // Capella
    // #[serde(with = "serde_utils::quoted_u64")]
    pub next_withdrawal_index: u64,
    // #[serde(with = "serde_utils::quoted_u64")]
    pub next_withdrawal_validator_index: u64,
    // Deep history valid from Capella onwards.
    pub historical_summaries: VariableList<HistoricalSummary, eth_spec::HistoricalRootsLimit>,
}

// TODO: Derive?
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct BeaconStatePrecomputedHashes {
    // Versioning
    pub genesis_time: Hash256,
    pub genesis_validators_root: Hash256,
    pub slot: Hash256,
    pub fork: Hash256,

    // History
    pub latest_block_header: Hash256,
    pub block_roots: Hash256,
    pub state_roots: Hash256,
    // Frozen in Capella, replaced by historical_summaries
    pub historical_roots: Hash256,

    // Ethereum 1.0 chain data
    pub eth1_data: Hash256,
    pub eth1_data_votes: Hash256,
    pub eth1_deposit_index: Hash256,

    // Registry
    pub validators: Hash256,
    pub balances: Hash256,

    // Randomness
    pub randao_mixes: Hash256,

    // Slashings
    pub slashings: Hash256,

    // Participation (Altair and later)
    pub previous_epoch_participation: Hash256,
    pub current_epoch_participation: Hash256,

    // Finality
    pub justification_bits: Hash256,
    pub previous_justified_checkpoint: Hash256,
    pub current_justified_checkpoint: Hash256,
    pub finalized_checkpoint: Hash256,

    // Inactivity
    pub inactivity_scores: Hash256,

    // Light-client sync committees
    pub current_sync_committee: Hash256,
    pub next_sync_committee: Hash256,

    // Execution
    pub latest_execution_payload_header: Hash256,

    // Capella
    pub next_withdrawal_index: Hash256,
    pub next_withdrawal_validator_index: Hash256,
    // Deep history valid from Capella onwards.
    pub historical_summaries: Hash256,
}

impl From<&BeaconState> for BeaconStatePrecomputedHashes {
    fn from(value: &BeaconState) -> Self {
        Self {
            genesis_time: value.genesis_time.tree_hash_root(),
            genesis_validators_root: value.genesis_validators_root.tree_hash_root(),
            slot: value.slot.tree_hash_root(),
            fork: value.fork.tree_hash_root(),
            latest_block_header: value.latest_block_header.tree_hash_root(),
            block_roots: value.block_roots.tree_hash_root(),
            state_roots: value.state_roots.tree_hash_root(),
            historical_roots: value.historical_roots.tree_hash_root(),
            eth1_data: value.eth1_data.tree_hash_root(),
            eth1_data_votes: value.eth1_data_votes.tree_hash_root(),
            eth1_deposit_index: value.eth1_deposit_index.tree_hash_root(),
            validators: value.validators.tree_hash_root(),
            balances: value.balances.tree_hash_root(),
            randao_mixes: value.randao_mixes.tree_hash_root(),
            slashings: value.slashings.tree_hash_root(),
            previous_epoch_participation: value.previous_epoch_participation.tree_hash_root(),
            current_epoch_participation: value.current_epoch_participation.tree_hash_root(),
            justification_bits: value.justification_bits.tree_hash_root(),
            previous_justified_checkpoint: value.previous_justified_checkpoint.tree_hash_root(),
            current_justified_checkpoint: value.current_justified_checkpoint.tree_hash_root(),
            finalized_checkpoint: value.finalized_checkpoint.tree_hash_root(),
            inactivity_scores: value.inactivity_scores.tree_hash_root(),
            current_sync_committee: value.current_sync_committee.tree_hash_root(),
            next_sync_committee: value.next_sync_committee.tree_hash_root(),
            latest_execution_payload_header: value.latest_execution_payload_header.tree_hash_root(),
            next_withdrawal_index: value.next_withdrawal_index.tree_hash_root(),
            next_withdrawal_validator_index: value.next_withdrawal_validator_index.tree_hash_root(),
            historical_summaries: value.historical_summaries.tree_hash_root(),
        }
    }
}

impl From<BeaconState> for BeaconStatePrecomputedHashes {
    fn from(value: BeaconState) -> Self {
        let borrowed: &BeaconState = &value;
        borrowed.into()
    }
}

// TODO: Derive?
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TreeHash)]
pub struct BeaconBlockHeaderPrecomputedHashes {
    pub slot: Hash256,
    pub proposer_index: Hash256,
    pub parent_root: Hash256,
    pub state_root: Hash256,
    pub body_root: Hash256,
}

impl From<&BeaconBlockHeader> for BeaconBlockHeaderPrecomputedHashes {
    fn from(value: &BeaconBlockHeader) -> Self {
        Self {
            slot: value.slot.tree_hash_root(),
            proposer_index: value.proposer_index.tree_hash_root(),
            parent_root: value.parent_root,
            state_root: value.state_root,
            body_root: value.body_root,
        }
    }
}

impl From<BeaconBlockHeader> for BeaconBlockHeaderPrecomputedHashes {
    fn from(value: BeaconBlockHeader) -> Self {
        let borrowed: &BeaconBlockHeader = &value;
        borrowed.into()
    }
}

File: synthetic_beacon_state_gen/eth_consensus_layer.py
from dataclasses import dataclass

from hexbytes import HexBytes
from ssz import get_hash_tree_root
from ssz.hash import hash_eth2
from ssz.hashable_container import HashableContainer
from ssz.sedes import (
    Bitlist,
    Bitvector,
    ByteVector,
    List,
    Vector,
    boolean,
    byte,
    bytes4,
    bytes32,
    bytes48,
    uint8,
    uint64,
    uint256
)

from eth_typing import HexStr
import constants
from typing import NewType, Dict, Any
from dataclasses import fields

Hash32 = bytes32
Root = bytes32
Gwei = uint64
Epoch = uint64
Slot = uint64
CommitteeIndex = uint64
ValidatorIndex = uint64
BLSPubkey = bytes48
ExecutionAddress = ByteVector(20)
WithdrawalIndex = uint64
ParticipationFlags = uint8
Version = bytes4

EpochNumber = NewType('EpochNumber', int)
FrameNumber = NewType('FrameNumber', int)
StateRoot = NewType('StateRoot', HexStr)
BlockRoot = NewType('BlockRoot', HexStr)
SlotNumber = NewType('SlotNumber', int)

BlockHash = NewType('BlockHash', HexStr)
BlockNumber = NewType('BlockNumber', int)


@dataclass
class FromResponse:
    """
    Class for extending dataclass with custom from_response method, ignored extra fields
    """

    @classmethod
    def from_response(cls, **kwargs) -> 'Self':
        class_field_names = [field.name for field in fields(cls)]
        return cls(**{k: v for k, v in kwargs.items() if k in class_field_names})

@dataclass
class BlockHeaderMessage(FromResponse):
    slot: str
    proposer_index: str
    parent_root: BlockRoot
    state_root: StateRoot
    body_root: str

class InclusionProofUtilsTrait:

    def get_merkle_tree_leafs(self):
        return self.hash_tree.chunks

    @property
    def _raw_hash_tree_for_inclusion(self):
        # Last layer only has one value that is the hash tree root
        return self.hash_tree.raw_hash_tree[:-1]

    def _get_field_index(self, field_name):
        return self._meta.field_names.index(field_name)

    def construct_inclusion_proof(self, field_name, field_hash):
        field_index = self._get_field_index(field_name)
        raw_hash_tree = self._raw_hash_tree_for_inclusion

        assert field_hash == raw_hash_tree[0][field_index],\
            (f"Field hash does not match the expected one.\n"
             f"Computed field index {field_index}.\n"
             f"Expected field hash: {raw_hash_tree[0][field_index]}.\n"
             f"Given field hash   :{field_hash}")

        result = []
        for layer in raw_hash_tree:
            sibling_idx = field_index - 1 if field_index % 2 == 1 else field_index + 1
            result.append(layer[sibling_idx])
            field_index //= 2
        return result

    def verify_inclusion_proof(self, field_name, field_hash, inclusion_proof):
        field_index = self._get_field_index(field_name)

        raw_hash_tree = self._raw_hash_tree_for_inclusion
        if len(inclusion_proof) != len(raw_hash_tree):
            return False

        current_hash = field_hash
        for idx in range(len(inclusion_proof)):
            inclusion_step, layer = inclusion_proof[idx], raw_hash_tree[idx]
            assert current_hash == layer[field_index], \
                (f"Verification failed at layer {idx}, field_index {field_index}.\n"
                 f"Expected{layer[field_index]}, got {current_hash}")
            if field_index % 2 == 0:
                current_hash = hash_eth2(current_hash + inclusion_step)
            else:
                current_hash = hash_eth2(inclusion_step + current_hash)

            field_index //= 2

        return current_hash == get_hash_tree_root(self)


class Fork(HashableContainer):
    fields = [
        ("previous_version", Version),
        ("current_version", Version),
        ("epoch", Epoch)  # Epoch of latest fork
    ]


class Checkpoint(HashableContainer):
    fields = [
        ("epoch", Epoch),
        ("root", Root)
    ]


class BeaconBlockHeader(HashableContainer, InclusionProofUtilsTrait):
    fields = [
        ("slot", Slot),
        ("proposer_index", ValidatorIndex),
        ("parent_root", Root),
        ("state_root", Root),
        ("body_root", Root),
    ]

    @classmethod
    def from_api(cls, api_response: BlockHeaderMessage) -> 'BeaconBlockHeader':
        return cls.create(
            slot = int(api_response.slot),
            proposer_index = int(api_response.proposer_index),
            parent_root = HexBytes(api_response.parent_root),
            state_root = HexBytes(api_response.state_root),
            body_root = HexBytes(api_response.body_root),
        )

    @classmethod
    def from_json(cls, json_response: Dict[str, Any]) -> 'BeaconBlockHeader':
        return cls.from_api(BlockHeaderMessage.from_response(**json_response))


class Eth1Data(HashableContainer):
    fields = [
        ("deposit_root", Root),
        ("deposit_count", uint64),
        ("block_hash", Hash32),
    ]


class Validator(HashableContainer):
    fields = [
        ("pubkey", BLSPubkey),
        ("withdrawal_credentials", bytes32),  # Commitment to pubkey for withdrawals
        ("effective_balance", Gwei),  # Balance at stake
        ("slashed", boolean),
        # Status epochs
        ("activation_eligibility_epoch", Epoch),  # When criteria for activation were met
        ("activation_epoch", Epoch),
        ("exit_epoch", Epoch),
        ("withdrawable_epoch", Epoch),  # When validator can withdraw funds
    ]


class AttestationData(HashableContainer):
    fields = [
        ("slot", Slot),
        ("index", CommitteeIndex),
        ("beacon_block_root", Root),
        ("source", Checkpoint),
        ("target", Checkpoint),
    ]


class PendingAttestation(HashableContainer):
    fields = [
        ("aggregation_bits", Bitlist(constants.MAX_VALIDATORS_PER_COMMITTEE)),
        ("data", AttestationData),
        ("inclusion_delay", Slot),
        ("proposer_index", ValidatorIndex),
    ]


class SyncCommittee(HashableContainer):
    fields = [
        ("pubkeys", Vector(BLSPubkey, constants.SYNC_COMMITTEE_SIZE)),
        ("aggregate_pubkey", BLSPubkey),
    ]


class ExecutionPayloadHeader(HashableContainer):
    # Execution block header fields
    fields = [
        ("parent_hash", Hash32),
        ("fee_recipient", ExecutionAddress),
        ("state_root", bytes32),
        ("receipts_root", bytes32),
        ("logs_bloom", ByteVector(constants.BYTES_PER_LOGS_BLOOM)),
        ("prev_randao", bytes32),
        ("block_number", uint64),
        ("gas_limit", uint64),
        ("gas_used", uint64),
        ("timestamp", uint64),
        # ("extra_data", ByteList(constants.MAX_EXTRA_DATA_BYTES)),
        # workaround - looks like ByteList is partially broken, but extra data is exactly bytes32
        ("extra_data", List(byte, constants.MAX_EXTRA_DATA_BYTES)),
        ("base_fee_per_gas", uint256),
        ("block_hash", Hash32),
        ("transactions_root", Root),
        ("withdrawals_root", Root),
        # Since Deneb
        ("blob_gas_used", uint64),
        ("excess_blob_gas", uint64),
    ]


class HistoricalSummary(HashableContainer):
    """
    `HistoricalSummary` matches the components of the phase0 `HistoricalBatch`
    making the two hash_tree_root-compatible.
    """
    fields = [
        ("block_summary_root", Root),
        ("state_summary_root", Root),
    ]


Validators = List(Validator, constants.VALIDATOR_REGISTRY_LIMIT)
Balances = List(Gwei, constants.VALIDATOR_REGISTRY_LIMIT)
JustificationBits = Bitvector(constants.JUSTIFICATION_BITS_LENGTH)


class BeaconState(HashableContainer, InclusionProofUtilsTrait):
    fields = [
        # Versioning
        ("genesis_time", uint64),
        ("genesis_validators_root", Root),
        ("slot", Slot),
        ("fork", Fork),
        # History
        ("latest_block_header", BeaconBlockHeader),
        ("block_roots", Vector(Root, constants.SLOTS_PER_HISTORICAL_ROOT)),
        ("state_roots", Vector(Root, constants.SLOTS_PER_HISTORICAL_ROOT)),
        ("historical_roots", List(Root, constants.HISTORICAL_ROOTS_LIMIT)),
        # Frozen in Capella, replaced by historical_summaries
        # Eth1
        ("eth1_data", Eth1Data),
        ("eth1_data_votes",
         List(Eth1Data, constants.EPOCHS_PER_ETH1_VOTING_PERIOD * constants.SLOTS_PER_EPOCH)),
        ("eth1_deposit_index", uint64),
        # Registry
        ("validators", Validators),
        ("balances", Balances),
        # Randomness
        ("randao_mixes", Vector(bytes32, constants.EPOCHS_PER_HISTORICAL_VECTOR)),
        # Slashings
        ("slashings", Vector(Gwei, constants.EPOCHS_PER_SLASHINGS_VECTOR)),
        # Per-epoch sums of slashed effective balances
        # Participation
        ("previous_epoch_participation", List(ParticipationFlags, constants.VALIDATOR_REGISTRY_LIMIT)),
        ("current_epoch_participation", List(ParticipationFlags, constants.VALIDATOR_REGISTRY_LIMIT)),
        # Finality
        ("justification_bits", JustificationBits),
        # Bit set for every recent justified epoch
        ("previous_justified_checkpoint", Checkpoint),
        ("current_justified_checkpoint", Checkpoint),
        ("finalized_checkpoint", Checkpoint),
        # Inactivity
        ("inactivity_scores", List(uint64, constants.VALIDATOR_REGISTRY_LIMIT)),
        # Sync
        ("current_sync_committee", SyncCommittee),
        ("next_sync_committee", SyncCommittee),
        # Execution
        ("latest_execution_payload_header", ExecutionPayloadHeader),  # (Modified in Capella)
        # Withdrawals
        ("next_withdrawal_index", WithdrawalIndex),  # (New in Capella)
        ("next_withdrawal_validator_index", ValidatorIndex),  # (New in Capella)
        # Deep history valid from Capella onwards
        ("historical_summaries", List(HistoricalSummary, constants.HISTORICAL_ROOTS_LIMIT)),
        # (New in Capella)
    ]
